server/s3_motr_rw_common.cc

```diff
@@ -89,7 +101,39 @@ void motr_op_done_on_main_thread(evutil_socket_t, short events,
              S3_IEM_MOTR_CONN_FAIL_JSON);
     }
     context->on_failed_handler()();  // Invoke the handler.
+
+    // Count ETIMEDOUT errors for any motr operations
+    // if number of errors for the etimedout_window sec
+    // greater than etimedout_max_threshold s3server should be
+    // restarted
+    if (error_code == -ETIMEDOUT && !gs_timeout_shutdown_in_progress) {
+
+      S3Option *optinst = S3Option::get_instance();
+      unsigned err_thr = optinst->get_motr_etimedout_max_threshold();
+      unsigned err_wnd = optinst->get_motr_etimedout_window_sec();
+      int64_t curtime = time(NULL);
+
+      if (curtime - gs_timeout_window_start >= (int64_t)err_wnd) {
+        s3_log(S3_LOG_DEBUG, request_id,
+               "Reset motr ETIMEDOUT window; cur window sec %" PRId64 ";",
+               curtime - gs_timeout_window_start);
+        gs_timeout_cnt = 0;
+      }
+      gs_timeout_cnt++;
+      gs_timeout_window_start = curtime;
+      if (gs_timeout_cnt >= err_thr) {
+        s3_log(
+            S3_LOG_ERROR, request_id,
+            "Shutdown. Motr ETIMEDOUT error cnt %u reached threshold value %u "
+            "for %" PRId64 " sec with allowed window %u sec",
+            gs_timeout_cnt, err_thr, curtime - gs_timeout_window_start,
+            err_wnd);
+        s3_kickoff_graceful_shutdown(0);
+        gs_timeout_shutdown_in_progress = true;
```

Comment:
should we set this value before calling graceful_shutdown() ?

Reply:
This flag indicates the function was called, so it is set after the call